File: engine/turingwars/src/main/scala/core/dto/GameUpdate.scala
package core.dto

case class GameUpdate(processes: List[Process], memory: List[MemoryUpdate], score: List[Score])

case class Process(processId: String, instructionPointer: Int, isAlive: Boolean)

case class MemoryUpdate(address: Int, cause: Int, value: Instruction)

case class Score(playerId: String, score: Int)

// TODO: Better definition for instruction. Maybe share with core
case class Instruction(op: String, a: InstructionField, b: InstructionField)

case class InstructionField(fieldType: String, value: Int, field: String)

File: web/src/shared/model/GameUpdate.ts
import { Instruction } from './Instruction';
import * as rt from 'runtypes';

export type Process = rt.Static<typeof Process>;
export const Process = rt.Record({
    processId: rt.String,
    instructionPointer: rt.Number,
    isAlive: rt.Boolean
});

export const MemoryUpdate = rt.Record({
    address: rt.Number,
    cause: rt.Number,
    value: Instruction
});

export const Score = rt.Record({
    playerId: rt.String,
    score: rt.Number
});

export type GameUpdate = rt.Static<typeof GameUpdate>;
export const GameUpdate = rt.Record({
    processes: rt.Array(Process),
    memory: rt.Array(MemoryUpdate),
    score: rt.Array(Score)
});